### PR TITLE
Convert first, last aggregate function to UDAF

### PR DESCRIPTION
**Pull request by mustafasrepo**
## Which issue does this PR close?

Closes #.

### CODE REVIEW

File: datafusion/functions/Cargo.toml [hunk 1]
```diff
@@ -80,7 +80,7 @@ itertools = { workspace = true }
 log = { workspace = true }
 md-5 = { version = "^0.10.0", optional = true }
 rand = { workspace = true }
-regex = { worksapce = true, optional = true }
+regex = { workspace = true, optional = true }
```
Comment (mustafasrepo) [inline datafusion/functions/Cargo.toml]: this was typo, fixed it on the way.

**Comment by mustafasrepo** [comment]
@ozankabak, this PR is ready for further review.

File: datafusion/physical-expr-common/src/aggregate/mod.rs [hunk 2]
```diff
@@ -303,6 +337,74 @@ impl AggregateExpr for AggregateFunctionExpr {
...
+                &reverse_udf,
+                &self.args,
+                &reverse_sort_exprs,
+                &reverse_ordering_req,
+                &self.schema,
+                name,
+                self.ignore_nulls,
+                self.is_distinct,
+            )
+            .unwrap();
```
Comment (ozankabak) [inline datafusion/physical-expr-common/src/aggregate/mod.rs:405]: Can we get rid of this `unwrap`?

File: datafusion/expr/src/udaf.rs [hunk 3]
```diff
@@ -193,6 +194,36 @@ impl AggregateUDF {
...
+        self.inner.order_sensitivity()
+    }
+
+    /// Reserves the `AggregateUDF` (e.g. returns the `AggregateUDF` that will
+    /// generate same result with this `AggregateUDF` when iterated in reverse
+    /// order, and `None` if there is no such `AggregateUDF`).
+    pub fn reverse_udf(&self) -> Option<AggregateUDF> {
+        match self.inner.reverse_expr() {
+            ReversedUDAF::NotSupported => None,
+            ReversedUDAF::Identical => Some(self.clone()),
```
Comment (ozankabak) [inline datafusion/expr/src/udaf.rs:219]: Can we get rid of this `clone`?

**Review by ozankabak — APPROVED** [review]
LGTM, two minor issues to investigate/discuss before we send it upstream.

**Comment by ozankabak** [comment]
Merged upstream.

**Pull request closed without merging**